File: src/TicTacToe.dart
import 'dart:io';

List<String> board = List<String>.filled(9, '');

void initializeBoard() {
  for (int i = 0; i < 9; i++) {
    board[i] = (i + 1).toString();
  }
}

void displayBoard() {
  print("\n--- Tic-Tac-Toe Board ---");
  print("-------------");
  for (int i = 0; i < 9; i += 3) {
    print("| ${board[i]} | ${board[i + 1]} | ${board[i + 2]} |");
    print("-------------");
  }
  print("-------------------------\n");
}

bool checkWin(String playerMark) {
  List<List<int>> winConditions = [
    [0, 1, 2], [3, 4, 5], [6, 7, 8],
    [0, 3, 6], [1, 4, 7], [2, 5, 8],
    [0, 4, 8], [2, 4, 6]
  ];
  for (var condition in winConditions) {
    if (board[condition[0]] == playerMark &&
        board[condition[1]] == playerMark &&
        board[condition[2]] == playerMark) {
      return true;
    }
  }
  return false;
}

bool checkDraw() {
  for (String cell in board) {
    if (int.tryParse(cell) != null) {
      return false;
    }
  }
  return true;
}

void main() {
  initializeBoard();
  String currentPlayer = 'X';
  bool gameOver = false;

  print("Welcome to Tic-Tac-Toe!");
  print("Players use 'X' and 'O'. Enter the number of the cell (1-9) to place your mark.");

  while (!gameOver) {
    displayBoard();

    int? choice;
    bool validMove = false;

    while (!validMove) {
      stdout.write("Player $currentPlayer's turn. Enter your move (1-9): ");
      String? input = stdin.readLineSync();

      if (input == null || input.isEmpty) {
        print("Invalid input. Please enter a number from 1 to 9.");
        continue;
      }

      choice = int.tryParse(input);

      if (choice != null) {
        int index = choice - 1;

        if (index >= 0 && index < 9 && int.tryParse(board[index]) != null) {
          board[index] = currentPlayer;
          validMove = true;
        } else if (index < 0 || index >= 9) {
          print("Invalid choice. Please enter a number from 1 to 9.");
        } else {
          print("Invalid move. That cell is already taken. Please choose an empty cell.");
        }
      } else {
        print("Invalid input. Please enter a number from 1 to 9.");
      }
    }

    if (checkWin(currentPlayer)) {
      displayBoard();
      print("Player $currentPlayer wins! Congratulations!");
      gameOver = true;
    } else if (checkDraw()) {
      displayBoard();
      print("It's a draw!");
      gameOver = true;
    } else {
      currentPlayer = (currentPlayer == 'X') ? 'O' : 'X';
    }
  }

  print("Game over. Thanks for playing!");
}

File: src/TicTacToe.rb
def initialize_board
  (1..9).to_a
end

def display_board(board)
  puts "\n--- Tic-Tac-Toe Board ---"
  puts "-------------"
  (0..2).each do |i|
    row_start_index = i * 3
    puts "| #{board[row_start_index]} | #{board[row_start_index + 1]} | #{board[row_start_index + 2]} |"
    puts "-------------"
  end
  puts "-------------------------\n"
end

def check_win(board, player_mark)
  win_conditions = [
    [0, 1, 2], [3, 4, 5], [6, 7, 8],
    [0, 3, 6], [1, 4, 7], [2, 5, 8],
    [0, 4, 8], [2, 4, 6]
  ]

  win_conditions.any? do |condition|
    board[condition[0]] == player_mark &&
      board[condition[1]] == player_mark &&
      board[condition[2]] == player_mark
  end
end

def check_draw(board)
  board.all? { |cell| cell == 'X' || cell == 'O' }
end

def play_game
  board = initialize_board
  current_player = 'X'
  game_over = false

  puts "Welcome to Tic-Tac-Toe!"
  puts "Players use 'X' and 'O'. Enter the number of the cell (1-9) to place your mark."

  until game_over
    display_board(board)

    valid_move = false
    choice = nil

    until valid_move
      print "Player #{current_player}'s turn. Enter your move (1-9): "
      input = gets.chomp

      if input =~ /^[1-9]$/
        choice = input.to_i
        index = choice - 1

        if board[index].is_a?(Integer)
          board[index] = current_player
          valid_move = true
        else
          puts "Invalid move. That cell is already taken. Please choose an empty cell."
        end
      else
        puts "Invalid input. Please enter a number from 1 to 9."
      end
    end

    if check_win(board, current_player)
      display_board(board)
      puts "Player #{current_player} wins! Congratulations!"
      game_over = true
    elsif check_draw(board)
      display_board(board)
      puts "It's a draw!"
      game_over = true
    else
      current_player = (current_player == 'X') ? 'O' : 'X'
    end
  end

  puts "Game over. Thanks for playing!"
end

play_game
